// jest-resolve/build/default_resolver.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = defaultResolver;

var _browserResolve;

function _load_browserResolve() {
  return _browserResolve = _interopRequireDefault(require('browser-resolve'));
}

var _fs;

function _load_fs() {
  return _fs = _interopRequireDefault(require('fs'));
}

var _path;

function _load_path() {
  return _path = _interopRequireDefault(require('path'));
}

var _is_builtin_module;

function _load_is_builtin_module() {
  return _is_builtin_module = _interopRequireDefault(require('./is_builtin_module'));
}

var _node_modules_paths;

function _load_node_modules_paths() {
  return _node_modules_paths = _interopRequireDefault(require('./node_modules_paths'));
}

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * 
 */

function defaultResolver(path, options) {
  const resolve = options.browser ? (_browserResolve || _load_browserResolve()).default.sync : resolveSync;

  return resolve(path, {
    basedir: options.basedir,
    extensions: options.extensions,
    moduleDirectory: options.moduleDirectory,
    paths: options.paths,
    rootDir: options.rootDir
  });
}

/*
 * Adapted from: https://github.com/substack/node-resolve
 */


const REGEX_RELATIVE_IMPORT = /^(?
